### PR TITLE
VP-6305: View existing customer segments

### PR DESCRIPTION
**Pull request by eMazeika**
Code includes all code copied from demo module. 
Feature to test: View existing customer segments
https://virtocommerce.atlassian.net/browse/VP-6305

### CODE REVIEW

File: src/VirtoCommerce.CustomerSegmentsModule.Core/Models/CustomerSegmentExpressionEvaluationContext.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace VirtoCommerce.CustomerSegmentsModule.Core.Models
+{
+    public class CustomerSegmentExpressionEvaluationContext : IEvaluationContext
+    {
+        public Contact Customer { get; set; }
+
+        public virtual bool CustomerHasPropertyValues(IEnumerable<DynamicObjectProperty> properties)
+        {
+            var result = properties.All(property =>
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CustomerSegmentsModule.Core/Models/CustomerSegmentExpressionEvaluationContext.cs]: Check for null before `properties.All`.
``` csharp
if (properties.IsNullOrEmpty())
    return false;
```

File: src/VirtoCommerce.CustomerSegmentsModule.Core/ModuleConstants.cs [hunk 2]
```diff
@@ -5,6 +5,9 @@ namespace VirtoCommerce.CustomerSegmentsModule.Core
 {
     public static class ModuleConstants
     {
+        public const int MaxAllowedSegments = 1000;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CustomerSegmentsModule.Core/ModuleConstants.cs]: Move it to settings.

File: src/VirtoCommerce.CustomerSegmentsModule.Data/Models/CustomerSegmentEntity.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        [StringLength(64)]
+        public string UserGroup { get; set; }
+
+        public bool IsActive { get; set; }
+
+        public DateTime? StartDate { get; set; }
+
+        public DateTime? EndDate { get; set; }
+
+        public string ExpressionTreeSerialized { get; set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CustomerSegmentsModule.Data/Models/CustomerSegmentEntity.cs]: ExpressionTreeSerialized looks like this is a boolean value. Can we rename it to the SerializedExpressionTree?

File: src/VirtoCommerce.CustomerSegmentsModule.Web/Controllers/Api/CustomerSegmentController.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            {
+                var result = AbstractTypeFactory<CustomerSegment>.TryCreateInstance();
+
+                result.ExpressionTree = AbstractTypeFactory<CustomerSegmentTree>.TryCreateInstance();
+                result.ExpressionTree.MergeFromPrototype(AbstractTypeFactory<CustomerSegmentTreePrototype>.TryCreateInstance());
+                result.IsActive = true;
+
+                return Ok(result);
+            }
+            else
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CustomerSegmentsModule.Web/Controllers/Api/CustomerSegmentController.cs]: No reason for `else` here.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.CustomerSegmentsModule.Data/Models/CustomerSegmentEntity.cs (hunk 3 above)
Comment (eMazeika) [reply src/VirtoCommerce.CustomerSegmentsModule.Data/Models/CustomerSegmentEntity.cs]: IsExpressionTreeSerialized is better for boolean (if it's boolean)

File: src/VirtoCommerce.CustomerSegmentsModule.Web/Scripts/directives/vaTagSelector.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                var totalCount = 0;
+
+                $scope.fetch = function ($select) {
+                    //loadEntityItems();
+                    if (!$scope.disabled) {
+                        $scope.fetchNext($select);
+                    }
+                };
+
+                //function loadEntityItems() {
```
Comment (AliveMen) [inline src/VirtoCommerce.CustomerSegmentsModule.Web/Scripts/directives/vaTagSelector.js]: Remove comments

**Review by AliveMen — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

**Review by akak1977 — APPROVED** [review]

**Pull request merged**